src/librustc/infer/error_reporting/nice_region_error/static_impl_trait.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright 2012-2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Error Reporting for static impl Traits.
+
+use infer::error_reporting::nice_region_error::NiceRegionError;
+use infer::lexical_region_resolve::RegionResolutionError;
+use ty::{BoundRegion, FreeRegion, RegionKind};
+use util::common::ErrorReported;
+
+impl<'a, 'gcx, 'tcx> NiceRegionError<'a, 'gcx, 'tcx> {
+    /// Print the error message for lifetime errors when the return type is a static impl Trait.
+    pub(super) fn try_report_static_impl_trait(&self) -> Option<ErrorReported> {
+        if let Some(ref error) = self.error {
+            match error.clone() {
+                RegionResolutionError::SubSupConflict(
+                    var_origin,
+                    sub_origin,
+                    sub_r,
+                    sup_origin,
+                    sup_r,
+                ) => {
+                    let anon_reg_sup = self.is_suitable_region(sup_r)?;
+                    if sub_r == &RegionKind::ReStatic &&
+                        self.is_return_type_impl_trait(anon_reg_sup.def_id)
```

Comment:
I think that it would be better if we could somehow detect that the `'static` was actually coming *from* the impl trait (as opposed to some other source). One way to do that might be to make use of the `sub_origin` variable -- but we'd have to update the origin that appears here:

https://github.com/rust-lang/rust/blob/b7c6e8f1805cd8a4b0a1c1f22f17a89e9e2cea23/src/librustc/infer/anon_types/mod.rs#L367

and maybe [here too](https://github.com/rust-lang/rust/blob/2a1c4eec40527de45b9d9b81672c8b9220d554fc/src/librustc/infer/anon_types/mod.rs#L770), though that could come later and is a bit more annoying (since that is an obligation cause). The current values are totally bogus anyway: they claim that `impl Trait` requirements arise "because returns types must be sized", which is just..not even kind of true.

Anyway, if we added a new `SubregionOrigin` for "impl trait obligation" and then checked here for that case, I think this would continue to work in the cases that we want, but rule out the cases we don't.

That said, I am not *that* keen on adding more uses of `SubregionOrigin`, which is "deprecated" (essentially) in the NLL checker. Still, I'm not sure what better thing to use for now.